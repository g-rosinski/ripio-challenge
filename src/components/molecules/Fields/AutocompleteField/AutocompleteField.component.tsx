import React, { ComponentProps } from "react"
import TextField from "../TextField/TextField.component"
import { ACContainer, ACOption, ACOptionsContainer, ACOptionsList } from "./AutocompleteField.styles"
import { useAutocomplete } from "../../../../hooks"

type AutocompleteFieldProps = ComponentProps<typeof TextField> & {
    onQuery: (query:string) => void,
    onSelectOption: (value:string) => void,
    options: string[]
}

const AutocompleteField: React.FC<AutocompleteFieldProps> = ({
    value,
    onChange,
    onSelectOption,
    onQuery,
    options,
    onFocus,
    onBlur,
    ...textFieldProps
}) => {
    const {
        displayOptions,
        handleOnChange,
        handleOnBlur,
        handleOnFocus,
        handleOnClickOption
    } = useAutocomplete<HTMLInputElement>({
        value,
        options,
        onQuery,
        onSelectOption,
        onChange,
        onFocus,
        onBlur
    })
    
    return (
        <ACContainer>
            <TextField 
                {...textFieldProps}
                value={value}
                onChange={handleOnChange}
                onFocus={handleOnFocus}
                onBlur={handleOnBlur}
                autoComplete="off"
            />
            { displayOptions && (
                <ACOptionsContainer>
                    <ACOptionsList role="list">
                        {options.map((option, i) => 
                            <ACOption 
                                key={i} 
                                role="listitem"
                                onClick={() => handleOnClickOption(option)} 
                                onMouseDown={e => e.preventDefault()}
                            >
                                {option}
                            </ACOption>
                        )}
                    </ACOptionsList>
                </ACOptionsContainer>
            )}
        </ACContainer>
    )
}

export default AutocompleteField