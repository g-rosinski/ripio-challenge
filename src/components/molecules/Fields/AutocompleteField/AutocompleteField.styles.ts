import styled from "styled-components";

export const ACContainer = styled.div`
    position: relative;
`

export const ACOptionsContainer = styled.div`
    position: absolute;
    bottom: 0;
    left: .7rem;
    right: .7rem;
    top: 80px;
    z-index: 3;
`

export const ACOptionsList = styled.ul`
    border-radius: 0.375rem;
    border: 1px solid #dee2e6;
    box-shadow: 1px 2px 5px #d0d0d0;
    padding: 0;

    & > li:nth-child(even){
        background-color: #dee2e6;

    }
`

export const ACOption = styled.li`
    background-color: #fff;
    list-style-type: none;
    padding: 0.375rem .7rem;
    cursor: pointer;

    &:hover{
        border-color: #86b7fe;
        outline: 0;
        box-shadow: 0 0 0 .25rem rgba(13,110,253,.25);
    }

    &:focus{
        border-color: #86b7fe;
        outline: 0;
        box-shadow: 0 0 0 .25rem rgba(13,110,253,.25);
    }

`