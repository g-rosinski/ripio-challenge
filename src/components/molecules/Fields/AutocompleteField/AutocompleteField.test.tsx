import { render, screen } from "@testing-library/react";
import React, { ComponentProps } from "react";
import AutocompleteField from "./AutocompleteField.component";

describe('<AutocompleteField />', () => {

    const onChange = jest.fn()
    const onQuery = jest.fn()
    const onSelectOption = jest.fn()
    const baseProps: ComponentProps<typeof AutocompleteField> = {
        label: 'Usuarios',
        name: 'users',
        value: '',
        options: [],
        onChange,
        onQuery,
        onSelectOption
    }

    test('Renderice correctamente', () => {
        const { container } = render(<AutocompleteField {...baseProps} />)

        expect(container).toMatchSnapshot()
        expect(screen.queryByRole('list')).not.toBeInTheDocument()
    })
})