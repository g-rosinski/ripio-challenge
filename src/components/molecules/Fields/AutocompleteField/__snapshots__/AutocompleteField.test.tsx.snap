// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteField /> Renderice correctamente 1`] = `
<div>
  <div
    class="sc-pyfCe gqyrsL"
  >
    <div
      class="sc-gswNZR kGxItT"
    >
      <label
        class="sc-dkrFOg hCrmYg"
        for="users"
      >
        Usuarios
      </label>
      <input
        autocomplete="off"
        class="sc-gKPRtg bujMUW"
        id="users"
        name="users"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
